src/rustdoc_gen.rs

```diff
@@ -123,11 +203,82 @@ pub(crate) enum CrateType<'a> {
     },
 }
 
+/// Configuration used to choose features to enable.
+/// Separate configs are used for baseline and current versions.
+#[derive(Debug, Clone, Hash)]
+pub(crate) struct FeatureConfig {
+    /// Feature set chosen as the foundation.
+    pub(crate) base_features: BaseFeatures,
+    /// Explicitly enabled features.
+    pub(crate) extra_features: Vec<String>,
+    pub(crate) is_baseline: bool,
+}
+
+#[derive(Debug, Clone, Hash)]
+pub(crate) enum BaseFeatures {
+    All,
+    Default,
+    Heuristic,
+    None,
+}
+
+impl FeatureConfig {
+    pub(crate) fn default_for_current() -> Self {
+        // The default behaviour for both version is the heuristic approach.
+        Self {
+            base_features: BaseFeatures::Heuristic,
+            extra_features: Vec::new(),
+            is_baseline: false,
+        }
+    }
+
+    pub(crate) fn default_for_baseline() -> Self {
+        Self {
+            base_features: BaseFeatures::Heuristic,
+            extra_features: Vec::new(),
+            is_baseline: true,
+        }
+    }
+
+    /// Unique identifier, used to mark generated rustdoc files.
+    /// `is_baseline` is ignored there, as rustdoc is cached only for baseline.
+    fn make_identifier(&self) -> String {
+        if matches!(self.base_features, BaseFeatures::Heuristic) && self.extra_features.is_empty() {
+            // If using the default config, return empty identifier
+            return String::new();
+        }
+
+        let mut extra_features = self.extra_features.clone();
+        // Sort the features, because the order is not revelant
+        extra_features.sort();
+
+        let unique_identifier = format!(
+            "{}{}",
+            match self.base_features {
+                BaseFeatures::All => "A",
+                BaseFeatures::Default => "D",
+                BaseFeatures::Heuristic => "H",
+                BaseFeatures::None => "N",
+            },
+            extra_features.join("@#$"), // separate features with some unusual symbols
+        );
+
+        // Calculate a rolling hash for the identifier, as it could end up too long.
+        let mut hash = Wrapping(0_u64);
+        let base = Wrapping(263_u64); // a hand-picked prime number
+        for b in unique_identifier.as_bytes() {
+            hash = hash * base + Wrapping(*b as u64);
+        }
+
+        format!("{:x}", hash)
+    }
```

Comment:
This sort of thing is where a design doc would have been useful again. It's okay to stop and discuss approaches instead of always trying to produce new code. I promise it's going to be faster and less frustrating to discuss possibly-tricky design decisions up front, instead of implementing them first and discussing them second. It also shows your maturity as an engineer when you flag possible tricky situations, instead of relying on me as a code reviewer to flag them for you.

Why are we building our own hash implementation inside `rustdoc_gen`? What is the benefit, what is the risk? What alternatives did you consider? Is this a good tradeoff?

This is a blocker.